src/main/kotlin/com/kotlindiscord/bot/filtering/Filter.kt

```diff
@@ -0,0 +1,106 @@
+package com.kotlindiscord.bot.filtering
+
+import com.gitlab.kordlib.core.behavior.channel.createMessage
+import com.gitlab.kordlib.core.entity.Message
+import com.gitlab.kordlib.core.entity.channel.TextChannel
+import com.gitlab.kordlib.core.event.message.MessageCreateEvent
+import com.gitlab.kordlib.core.event.message.MessageUpdateEvent
+import com.gitlab.kordlib.rest.builder.message.MessageCreateBuilder
+import com.gitlab.kordlib.rest.request.RequestException
+import com.kotlindiscord.bot.config.config
+import com.kotlindiscord.bot.enums.Channels
+import com.kotlindiscord.bot.enums.Roles
+import com.kotlindiscord.kord.extensions.ExtensibleBot
+import io.ktor.http.HttpStatusCode
+
+/**
+ * Class representing a single type of message filter.
+ *
+ * Subclasses are expected to implement exactly one type of message filter, and they must implement
+ * the [checkCreate] function. This function should check whether a message is problematic, action it
+ * appropriately, and return `false` if processing should stop here, or `true` if other filters
+ * should be checked.
+ *
+ * @param bot Current bot instance
+ */
+abstract class Filter(val bot: ExtensibleBot) {
+    /** Message appended to notification reasons sent to users. **/
+    val mistakeMessage = "If you feel that this was a mistake, please feel free to contact a member of staff."
+
+    /**
+     * An array of [FilterConcerns] - filters won't be executed if their concerns aren't met.
+     *
+     * This means that we don't waste time executing filters when the data they filter against isn't present.
+     **/
+    abstract val concerns: Array<FilterConcerns>
+
+    /**
+     * Check whether a message should be actioned, action it, and return a value based on
+     * whether we should continue processing filters.
+     *
+     * This function is only executed on message creation.
```

Comment:
You should add a note here about filters that only send an alert but don't delete the message should return true even if the message triggered the filter.

Reply:
I'm not sure if that's necessarily always going to be the case - that's just how the filters are set up currently. Is it worth enforcing?

Reply:
Maybe a more general message would be better, but I think we should still put a note about it, at least for now.